sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/ColumnarTestUtils.scala

```diff
@@ -54,12 +54,22 @@ object ColumnarTestUtils {
       case COMPACT_DECIMAL(precision, scale) => Decimal(Random.nextLong() % 100, precision, scale)
       case LARGE_DECIMAL(precision, scale) => Decimal(Random.nextLong(), precision, scale)
       case STRUCT(_) =>
-        new GenericInternalRow(Array[Any](UTF8String.fromString(Random.nextString(10))))
+        val schema = StructType(Array(StructField("test", StringType)))
+        val converter = UnsafeProjection.create(schema)
+        converter(InternalRow(Array(UTF8String.fromString(Random.nextString(10))): _*))
       case ARRAY(_) =>
-        new GenericArrayData(Array[Any](Random.nextInt(), Random.nextInt()))
+        UnsafeArrayData.fromPrimitiveArray(Array(Random.nextInt(), Random.nextInt()))
       case MAP(_) =>
-        ArrayBasedMapData(
-          Map(Random.nextInt() -> UTF8String.fromString(Random.nextString(Random.nextInt(32)))))
+        val unsafeConverter =
+          UnsafeProjection.create(Array[DataType](MapType(IntegerType, StringType)))
+        val row = new GenericInternalRow(1)
+        def toUnsafeMap(map: ArrayBasedMapData): UnsafeMapData = {
+          row.update(0, map)
+          val unsafeRow = unsafeConverter.apply(row)
+          unsafeRow.getMap(0).copy
+        }
+        toUnsafeMap(ArrayBasedMapData(
+          Map(Random.nextInt() -> UTF8String.fromString(Random.nextString(Random.nextInt(32))))))
```

Comment:
Seems above changes to data generation are unnecessary too?

Reply:
The ColumnType for Maps/Struct/Array all end up casting to their Unsafe structures to get the size for the statistics, so the test data will need to reflect that as well.